src/cfn-resource-generator.ts

```diff
@@ -0,0 +1,144 @@
+    this.schema.readOnlyProperties.map((prop: string) => prop.replace(/^\/properties\//, '')).forEach((prop: string) => {
+      delete schema.properties[prop];
+    });
+
+    gen.emitType(`Cfn${this.sanitizedTypeName}Props`, schema);
+
+    gen.renderToCode(code);
+    code.line();
+  }
+
+  private generateConstructClass(code: j2j.Code) {
+    code.line('/**');
+    code.line(` * A CloudFormation \`${this.typeName}\``);
+    code.line(' *');
+    code.line(` * @cloudformationResource ${this.typeName}`);
+    code.line(' * @stability external');
+    code.line(` * @link ${this.typeDef.SourceUrl}`);
+    code.line(' */');
+    code.openBlock(`export class Cfn${this.sanitizedTypeName} extends cdk.CfnResource`);
+    code.line('/**');
+    code.line('* The CloudFormation resource type name for this resource class.');
+    code.line('*/');
+    code.line(`public static readonly CFN_RESOURCE_TYPE_NAME = "${this.typeName}";`);
+    code.line();
+
+    Object.keys(this.schema.properties).filter(prop => this.schema.readOnlyProperties.indexOf(`/properties/${prop}`) < 0).forEach(prop => {
+      const optionalMarker = this.schema.required.indexOf(prop) < 0 ? ' | undefined' : '';
+      code.line('/**');
+      code.line(` * \`${this.typeName}.${prop}\``);
+      if (this.schema.properties[prop].description) {
+        code.line(` * ${this.schema.properties[prop].description}`);
+      }
+      code.line(` * @link ${this.typeDef.SourceUrl}`);
+      code.line(' */');
+      code.line(`public readonly ${camelcase(prop)}: ${this.getTypeOfProperty(prop)}${optionalMarker};`);
+    });
+    Object.keys(this.schema.properties).filter(prop => this.schema.readOnlyProperties.indexOf(`/properties/${prop}`) >= 0).forEach(prop => {
+      code.line('/**');
+      code.line(` * Attribute \`${this.typeName}.${prop}\``);
+      code.line(` * @link ${this.typeDef.SourceUrl}`);
+      code.line(' */');
+      code.line(`public readonly attr${camelcase(prop, { pascalCase: true })}: ${this.getTypeOfProperty(prop)};`);
+    });
+
+    code.line();
+
+    code.line('/**');
+    code.line(` * Create a new \`${this.typeName}\`.`);
+    code.line(' *');
+    code.line(' * @param scope - scope in which this resource is defined');
+    code.line(' * @param id    - scoped id of the resource');
+    code.line(' * @param props - resource properties');
+    code.line(' */');
+    code.openBlock(`constructor(scope: cdk.Construct, id: string, props: Cfn${this.sanitizedTypeName}Props)`);
+    code.line(`super(scope, id, { type: Cfn${this.sanitizedTypeName}.CFN_RESOURCE_TYPE_NAME, properties: toJson_Cfn${this.sanitizedTypeName}Props(props)! });`);
+    this.schema.required.forEach((prop: string) => {
+      code.line(`cdk.requireProperty(props, \'${camelcase(prop)}\', this);`);
+    });
+    code.line('');
+    Object.keys(this.schema.properties).filter(prop => this.schema.readOnlyProperties.indexOf(`/properties/${prop}`) < 0).forEach(prop => {
```

Comment:
This is what I meant when I said to pre-process and store as private members. Then you don't need to duplicate the logic that separates properties and attributes everywhere.

```suggestion
    this.resourceProperties.forEach(prop => {
```